### PR TITLE
[MSPAINT][ATL] Encapsulation: mainWindow

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Improve source code quality.
JIRA issue: [CORE-18867](https://jira.reactos.org/browse/CORE-18867)

## Proposed changes

- Add `CFullscreenWindow::DoCreate`, `CMiniatureWindow::DoCreate`, and `CMainWindow::DoCreate` helper functions.
- Encapsulation around `mainWindow` and `_tWinMain`.
- Add `CMainWindow::GetOpenFileName`, `CMainWindow::GetSaveFileName`, and `CMainWindow::ChooseColor` helper functions (optimized for startup).
- Move some code in `WinMain` into `CMainWindow::OnCreate`.
- Delay creation of `CFullscreenWindow` and `CMiniatureWindow` (optimized for startup and memory usage).
- Add `atlimagedx.h` in `mspaint`.
- Extend ATL `CImage` class as `CImageDx` in `mspaint`.

## Comparison

Memory usage (BEFORE):
before

Memory usage (AFTER):
after

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by katahiromz** [comment]
It's go.

File: base/applications/mspaint/dib.cpp [hunk 1]
```diff
@@ -6,11 +6,14 @@
...
  */
 
-/* INCLUDES *********************************************************/
-
 #include "precomp.h"
 #include <math.h>
 
+INT fileSize = 0;
+INT fileHPPM = 2834; // horizontal pixels per millimeter (PPM)
+INT fileVPPM = 2834; // vertical pixels per millimeter (PPM)
```
Comment (HBelusca) [inline base/applications/mspaint/dib.cpp:14]: Where do these value come from?

File: base/applications/mspaint/main.cpp [hunk 2]
```diff
@@ -115,193 +82,165 @@ OFNHookProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
...
+        CString strAllPictureFiles;
+        strAllPictureFiles.LoadString(hProgInstance, IDS_ALLPICTUREFILES);
+
+        // Get the import filter
+        CSimpleArray<GUID> aguidFileTypesI;
+        CImage::GetImporterFilterString(strFilter, aguidFileTypesI, strAllPictureFiles,
+                                        CImage::excludeDefaultLoad, _T('\0'));
+
+        // Initializing the OPENFILENAME structure for GetOpenFileName
+        ZeroMemory(&ofn, sizeof(OPENFILENAME));
```
Comment (HBelusca) [inline base/applications/mspaint/main.cpp:102]: You can use (for these sizeofs): `sizeof(ofn)` instead.

File: base/applications/mspaint/main.cpp [hunk 3]
```diff
@@ -115,193 +82,165 @@ OFNHookProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
...
+
+    if (sfn.lStructSize == 0)
+    {
+        // Get the export filter
+        CSimpleArray<GUID> aguidFileTypesE;
+        CImage::GetExporterFilterString(strFilter, aguidFileTypesE, NULL,
+                                        CImage::excludeDefaultSave, _T('\0'));
+
+        // Initializing the OPENFILENAME structure for GetSaveFileName
+        ZeroMemory(&sfn, sizeof(OPENFILENAME));
```
Comment (HBelusca) [inline base/applications/mspaint/main.cpp:129]: Same

File: base/applications/mspaint/main.cpp [hunk 4]
```diff
@@ -115,193 +82,165 @@ OFNHookProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
...
+    {
         0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff,
         0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff
     };
 
+    if (choosecolor.lStructSize == 0)
+    {
+        // Initializing the CHOOSECOLOR structure for ChooseColor
+        ZeroMemory(&choosecolor, sizeof(choosecolor));
+        choosecolor.lStructSize    = sizeof(CHOOSECOLOR);
```
Comment (HBelusca) [inline base/applications/mspaint/main.cpp:167]:
```suggestion
        choosecolor.lStructSize    = sizeof(choosecolor);
```

File: base/applications/mspaint/winproc.cpp [hunk 5]
```diff
@@ -252,8 +254,51 @@ LRESULT CMainWindow::OnDropFiles(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL&
...
+    {
+        miniature.DoCreate(m_hWnd);
+        miniature.ShowWindow(SW_SHOWNOACTIVATE);
+    }
+
+    // Initialize imageModel
+    imageModel.Crop(registrySettings.BMPWidth, registrySettings.BMPHeight);
+    if (__argc >= 2)
+        DoLoadImageFile(m_hWnd, __targv[1], TRUE);
+    imageModel.ClearHistory();
```
Comment (HBelusca) [inline base/applications/mspaint/winproc.cpp:296]: Wouldn't it be more meaningful to do that DoLoadImageFile call (and maybe the clearhistory call as well) **within** the WinMain function, once after you have created the `mainWindow` object ?
(Even if there is so far only one instance of the mainwindow, this would separate the general object from how it's going to be used next, namely, that we are going to load something in it.)

File: base/applications/mspaint/main.cpp (hunk 2 above)
Comment (katahiromz) [reply base/applications/mspaint/main.cpp]: Fixed.

File: base/applications/mspaint/main.cpp (hunk 3 above)
Comment (katahiromz) [reply base/applications/mspaint/main.cpp]: Fixed.

File: base/applications/mspaint/main.cpp (hunk 4 above)
Comment (katahiromz) [reply base/applications/mspaint/main.cpp]: Fixed.

File: base/applications/mspaint/dib.cpp (hunk 1 above)
Comment (katahiromz) [reply base/applications/mspaint/dib.cpp:14]: Win2k3 mspaint:
image
101 dots per inch.

ReactOS mspaint on Win10:
image
2834 pixels per meter. This value is wrong.

I have asked ChatGPT for it:

```txt
101 dots per inch = 101 dots / 1 inch
= 101 dots / 0.0254 meters (1 inch = 0.0254 meters)
= 3980.31 dots per meter
```

So we have to fix the `fileHPPM` and `fileVPPM` values.

```txt
1 dots per inch = 1 dots / 1 inch
= 1 dots / 0.0254 meters (1 inch = 0.0254 meters)
= 39.37 dots per meter
```

File: base/applications/mspaint/winproc.cpp (hunk 5 above)
Comment (katahiromz) [reply base/applications/mspaint/winproc.cpp]: Moved them into `WinMain`.

File: base/applications/mspaint/dib.cpp (hunk 1 above)
Comment (katahiromz) [reply base/applications/mspaint/dib.cpp:14]: `ATL` `CImage` class has no method to get the image resolution.
`gdiplus` `Image` class has `GetHorizontalResolution` and `GetVerticalResolution` methods.

Comment (katahiromz) [reply base/applications/mspaint/dib.cpp]: Now fixed.

File: sdk/lib/atl/atlimage.h [hunk 6]
```diff
@@ -374,7 +374,11 @@ class CImage
         return m_hbm == NULL;
     }
 
+#ifdef MSPAINT
+    HRESULT Load(LPCTSTR pszFileName, float *pxDpi = NULL, float *pyDpi = NULL) throw()
```
Comment (HBelusca) [inline sdk/lib/atl/atlimage.h:378]: No, please revert those ATL changes, as you are breaking compatibility with MS.
Either you are able to retrieve the information needed with the existing functions defined there
(or add any missing function that also exist in the official ATL, see https://learn.microsoft.com/en-us/cpp/atl-mfc-shared/reference/cimage-class?view=msvc-170 ), or, you call those extra functions e.g. the GetImageHorizontalResolution() in mspaint itself to retrieve the missing information. Either directly (just after you call the Load/Save methods of CImage in mspaint), or by defining a class derived from CImage, within mspaint, that adds the extra functionality.

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: sdk/lib/atl/atlimage.h (hunk 6 above)
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: I added `CImageDx` that extends `CImage`.
Fixed.

File: base/applications/mspaint/winproc.cpp [hunk 7]
```diff
@@ -640,24 +690,35 @@ LRESULT CMainWindow::OnCommand(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL& bH
             break;
         }
         case IDM_EDITCOPYTO:
-            if (GetSaveFileName(&ofn))
-                SaveDIBToFile(selectionModel.GetBitmap(), ofn.lpstrFile, imageModel.GetDC());
+        {
+            TCHAR szFileName[MAX_LONG_PATH];
+            szFileName[0] = 0;
```
Comment (HBelusca) [inline base/applications/mspaint/winproc.cpp:695]:
```suggestion
            TCHAR szFileName[MAX_LONG_PATH] = _T("");
```

File: base/applications/mspaint/winproc.cpp [hunk 8]
```diff
@@ -640,24 +690,35 @@ LRESULT CMainWindow::OnCommand(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL& bH
...
+            if (DoGetSaveFileName(szFileName, _countof(szFileName)))
+                SaveDIBToFile(selectionModel.GetBitmap(), szFileName, imageModel.GetDC());
             break;
+        }
         case IDM_EDITPASTEFROM:
-            if (GetOpenFileName(&ofn))
+        {
+            TCHAR szFileName[MAX_LONG_PATH];
+            szFileName[0] = 0;
```
Comment (HBelusca) [inline base/applications/mspaint/winproc.cpp:703]:
```suggestion
            TCHAR szFileName[MAX_LONG_PATH] = _T("");
```

File: base/applications/mspaint/winproc.cpp [hunk 9]
```diff
@@ -496,11 +542,15 @@ LRESULT CMainWindow::OnCommand(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL& bH
             }
             break;
         case IDM_FILEOPEN:
-            if (ConfirmSave() && GetOpenFileName(&ofn))
             {
-                DoLoadImageFile(m_hWnd, ofn.lpstrFile, TRUE);
+                TCHAR szFileName[MAX_LONG_PATH];
+                szFileName[0] = 0;
```
Comment (HBelusca) [inline base/applications/mspaint/winproc.cpp:547]:
```suggestion
                TCHAR szFileName[MAX_LONG_PATH] = _T("");
```

File: base/applications/mspaint/main.cpp [hunk 10]
```diff
@@ -115,193 +82,171 @@ OFNHookProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
...
+    {
         0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff,
         0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff, 0xffffff
     };
 
+    if (choosecolor.lStructSize == 0)
+    {
+        // Initializing the CHOOSECOLOR structure for ChooseColor
+        ZeroMemory(&choosecolor, sizeof(choosecolor));
+        choosecolor.lStructSize    = sizeof(choosecolor);
```
Comment (HBelusca) [inline base/applications/mspaint/main.cpp:167]: you can remove the extra spaces at the left of the equal sign (for those three lines).

File: base/applications/mspaint/main.cpp [hunk 11]
```diff
@@ -115,193 +82,171 @@ OFNHookProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
...
+    if (sfn.lStructSize == 0)
+    {
+        // Get the export filter
+        CSimpleArray<GUID> aguidFileTypesE;
+        CImage::GetExporterFilterString(strFilter, aguidFileTypesE, NULL,
+                                        CImage::excludeDefaultSave, _T('\0'));
+
+        // Initializing the OPENFILENAME structure for GetSaveFileName
+        ZeroMemory(&sfn, sizeof(sfn));
+        sfn.lStructSize    = sizeof(sfn);
```
Comment (HBelusca) [inline base/applications/mspaint/main.cpp:130]: please remove the extra spaces at the left of the equals signs.

File: base/applications/mspaint/main.cpp [hunk 12]
```diff
@@ -115,193 +82,171 @@ OFNHookProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
...
+        CString strAllPictureFiles;
+        strAllPictureFiles.LoadString(hProgInstance, IDS_ALLPICTUREFILES);
+
+        // Get the import filter
+        CSimpleArray<GUID> aguidFileTypesI;
+        CImage::GetImporterFilterString(strFilter, aguidFileTypesI, strAllPictureFiles,
+                                        CImage::excludeDefaultLoad, _T('\0'));
+
+        // Initializing the OPENFILENAME structure for GetOpenFileName
+        ZeroMemory(&ofn, sizeof(ofn));
```
Comment (HBelusca) [inline base/applications/mspaint/main.cpp:102]: remove extra spaces at the left of equals signs.

File: base/applications/mspaint/dialogs.cpp [hunk 13]
```diff
@@ -162,18 +164,19 @@ LRESULT CAttributesDialog::OnRadioButton3(WORD wNotifyCode, WORD wID, HWND hWndC
 
 LRESULT CAttributesDialog::OnEdit1(WORD wNotifyCode, WORD wID, HWND hWndCtl, BOOL& bHandled)
 {
+    FixDpi();
```
Comment (HBelusca) [inline base/applications/mspaint/dialogs.cpp:167]: Can't you just call this function at the only place where those `g_xDpi` and `g_yDpi` are (re)initialized at runtime?

File: base/applications/mspaint/atlimagedx.h [hunk 14]
```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+class CImageDx : public CImage
```
Comment (HBelusca) [inline base/applications/mspaint/atlimagedx.h:10]: What does "Dx" mean here?

File: base/applications/mspaint/atlimagedx.h [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        BitmapSetResolution(pBitmap, xDpi, yDpi);
+        return TRUE;
+    }
+
+    HRESULT LoadDx(LPCTSTR pszFileName, float *pxDpi, float *pyDpi) throw()
+    {
+        // convert the file name string into Unicode
+        CStringW pszNameW(pszFileName);
+
+        // create a GpBitmap object from file
```
Comment (HBelusca) [inline base/applications/mspaint/atlimagedx.h:63]: Why not just calling CImage::Load instead of seemingly redoing the whole work?

File: base/applications/mspaint/dialogs.cpp [hunk 16]
```diff
@@ -135,20 +135,22 @@ LRESULT CAttributesDialog::OnDefault(WORD wNotifyCode, WORD wID, HWND hWndCtl, B
...
 
 LRESULT CAttributesDialog::OnRadioButton2(WORD wNotifyCode, WORD wID, HWND hWndCtl, BOOL& bHandled)
 {
+    FixDpi();
     CString strNum;
-    strNum.Format(_T("%.3lf"), newWidth * 100.0 / fileHPPM);
+    strNum.Format(_T("%.3lf"), newWidth / (PpmFromDpi(g_xDpi) / 100));
     SetDlgItemText(IDD_ATTRIBUTESEDIT1, strNum);
-    strNum.Format(_T("%.3lf"), newHeight * 100.0 / fileVPPM);
+    strNum.Format(_T("%.3lf"), newHeight / (PpmFromDpi(g_yDpi) / 100));
```
Comment (HBelusca) [inline base/applications/mspaint/dialogs.cpp:153]: Won't this way of doing the division create rounding errors?

File: base/applications/mspaint/atlimagedx.h [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        GetImageVerticalResolution(pImage, pyDpi);
+        return TRUE;
+    }
+
+    BOOL SetResolution(Gdiplus::GpBitmap *pBitmap, float xDpi, float yDpi) const
+    {
+        if (xDpi <= 0)
+            xDpi = 96;
+        if (yDpi <= 0)
+            yDpi = 96;
```
Comment (HBelusca) [inline base/applications/mspaint/atlimagedx.h:46]: Why not just keeping whatever is given to the function? (Meaning also that your default values in SaveDx may need to be 96 instead.)

File: base/applications/mspaint/fullscreen.cpp [hunk 18]
```diff
@@ -6,12 +6,21 @@
...
 
-/* INCLUDES *********************************************************/
-
 #include "precomp.h"
 
+CFullscreenWindow fullscreenWindow;
+
 /* FUNCTIONS ********************************************************/
 
+HWND CFullscreenWindow::DoCreate()
```
Comment (HBelusca) [inline base/applications/mspaint/fullscreen.cpp:15]: Why not calling all those "DoCreate" functions, by "Create" instead?

File: base/applications/mspaint/winproc.h [hunk 19]
```diff
@@ -28,7 +28,16 @@ class CMainWindow : public CWindowImpl<CMainWindow>
         MESSAGE_HANDLER(WM_MOUSEWHEEL, OnMouseWheel)
     END_MSG_MAP()
 
+    CMainWindow() : m_hMenu(NULL) { }
+
+    HWND DoCreate();
+    BOOL DoGetOpenFileName(IN OUT LPTSTR pszFile, INT cchMaxFile);
+    BOOL DoGetSaveFileName(IN OUT LPTSTR pszFile, INT cchMaxFile);
+    BOOL DoChooseColor(IN OUT COLORREF *prgbColor);
```
Comment (HBelusca) [inline base/applications/mspaint/winproc.h:36]: I'm not convinced by the need to add "Do" to all those function names.

File: sdk/lib/atl/atlimage.h [hunk 20]
```diff
@@ -828,20 +829,20 @@ class CImage
...
+    // abbreviations of GDI+ basic types
+    typedef Gdiplus::GpStatus St;
+    typedef Gdiplus::ImageCodecInfo ICI;
+    typedef Gdiplus::GpBitmap Bm;
+    typedef Gdiplus::EncoderParameters EncParams;
+    typedef Gdiplus::GpImage Im;
+    typedef Gdiplus::ARGB ARGB;
+    typedef HBITMAP HBM;
+    typedef Gdiplus::GdiplusStartupInput GSI;
+    typedef Gdiplus::GdiplusStartupOutput GSO;
```
Comment (HBelusca) [inline sdk/lib/atl/atlimage.h:841]: Is this change necessary?

File: base/applications/mspaint/main.cpp (hunk 10 above)
Comment (katahiromz) [reply base/applications/mspaint/main.cpp]: Removed.

File: base/applications/mspaint/main.cpp (hunk 11 above)
Comment (katahiromz) [reply base/applications/mspaint/main.cpp]: Removed.

File: base/applications/mspaint/main.cpp (hunk 12 above)
Comment (katahiromz) [reply base/applications/mspaint/main.cpp:102]: Removed.

File: base/applications/mspaint/atlimagedx.h (hunk 14 above)
Comment (katahiromz) [reply base/applications/mspaint/atlimagedx.h:10]: "dirty extension"

File: base/applications/mspaint/atlimagedx.h (hunk 15 above)
Comment (katahiromz) [reply base/applications/mspaint/atlimagedx.h:63]: `GpBitmap` is not available from outside of `CImage::Load`.

File: base/applications/mspaint/dialogs.cpp (hunk 16 above)
Comment (katahiromz) [reply base/applications/mspaint/dialogs.cpp:153]: The return value of `PpmFromDpi` is a `float`.

File: base/applications/mspaint/dialogs.cpp (hunk 13 above)
Comment (katahiromz) [reply base/applications/mspaint/dialogs.cpp]: `FixDpi()` is deleted.

File: base/applications/mspaint/atlimagedx.h (hunk 17 above)
Comment (katahiromz) [reply base/applications/mspaint/atlimagedx.h]: Deleted.

File: base/applications/mspaint/fullscreen.cpp (hunk 18 above)
Comment (katahiromz) [reply base/applications/mspaint/fullscreen.cpp:15]: The super class has already `Create(...)` method.

File: base/applications/mspaint/dialogs.cpp (hunk 16 above)
Comment (katahiromz) [reply base/applications/mspaint/dialogs.cpp]: Improved.

File: base/applications/mspaint/fullscreen.cpp (hunk 18 above)
Comment (HBelusca) [reply base/applications/mspaint/fullscreen.cpp:15]: But which `Create(...)` ? Create without parameters?

File: sdk/lib/atl/atlimage.h (hunk 20 above)
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: Reduced changes.

File: base/applications/mspaint/fullscreen.cpp (hunk 18 above)
Comment (katahiromz) [reply base/applications/mspaint/fullscreen.cpp:15]: `CWindowImplBaseT<CWindow, CControlWinTraits>::Create`.

File: base/applications/mspaint/winproc.h (hunk 19 above)
Comment (katahiromz) [reply base/applications/mspaint/winproc.h]: Reduced 'Do's.

File: base/applications/mspaint/fullscreen.cpp (hunk 18 above)
Comment (katahiromz) [reply base/applications/mspaint/fullscreen.cpp:15]: This is a template class. Overriding is difficult.
